### PR TITLE
read/macho: support Go's compression format

### PR DESCRIPTION
**Pull request by ajwerner**
This PR adds support in the object crate for Go's format to compress Mach-O debug info sections. It contains two changes:

* The MachOFile now can match section names which contain both the `__` and `__z` prefix instead of a `.`.
* The MachOSection now detects the presence of the `__z` prefix and attempts to offer decompression when it detects the section is compressed using the Go compression scheme.

A third commit adds testing. This third commit will need to be rebased to update the testfiles commit once https://github.com/gimli-rs/object-testfiles/pull/14 is merged into the upstream.

Fixes https://github.com/gimli-rs/object/issues/696.

### CODE REVIEW

File: src/read/macho/section.rs [hunk 1]
```diff
@@ -102,6 +102,48 @@ where
             .data(self.file.endian, self.internal.data)
             .read_error("Invalid Mach-O section size or offset")
     }
+
+    // Try Go-style "ZLIB" header decompression. This is similar to the GNU-style
+    // format, but the header uses a slightly different format with a 4-byte "ZLIB"
+    // magic followed by a 64-bit big-endiang uncompressed size.
```
Comment (bjorn3) [inline src/read/macho/section.rs:108]:
```suggestion
    // magic followed by a 64-bit big-endian uncompressed size.
```

Comment (ajwerner) [reply src/read/macho/section.rs]: Done.

**Comment by ajwerner** [comment]
TFTR!

I also manually tried this out with gimli's `dwarfdump` and it works. 

Before:
```
$  cargo run --bin dwarfdump -- ../object-testfiles/macho/go-x86_64
.debug_info

.debug_types

.debug_pubnames

.debug_aranges

.debug_pubtypes
```

After:
```
$ git diff
diff --git a/Cargo.toml b/Cargo.toml
index 8eed4b2..491566f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
diff --git a/Cargo.toml b/Cargo.toml
index 8eed4b2..05faeb6 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -59,3 +59,6 @@ split-debuginfo = "packed"
 members = ["crates/examples"]
 default-members = [".", "crates/examples"]
 resolver = "2"
+
+[patch.crates-io]
+object = { path = "/Users/ajwerner/src/github.com/gimli-rs/object" }
```

```
$ cargo run --bin dwarfdump -- ../object-testfiles/macho/go-x86_64


.debug_info

UNIT<.debug_info+0x00000000>: length = 0xc99, format = Dwarf32, version = 4, address_size = 8, abbrev_offset = 0x0
< 0><0x0000000b>  DW_TAG_compile_unit
                    DW_AT_name                  internal/fmtsort
                    DW_AT_language              DW_LANG_Go
                    DW_AT_stmt_list             <.debug_line+0x00000000>
                    DW_AT_low_pc                0x01074de0
                    DW_AT_ranges                <rnglist at .debug_ranges+0x00000000>
                        [ 0] <offset-pair 0x00000000, 0x0000178c> [0x01074de0, 0x0107656c]
                    DW_AT_comp_dir              .
                    DW_AT_producer              Go cmd/compile go1.22.4; -shared regabi
                    Unknown DwAt: 10501         fmtsort
< 1><0x00000060>    DW_TAG_subprogram
                      DW_AT_name                  reflect.Value.MapRange
                      DW_AT_inline                DW_INL_inlined
                      DW_AT_decl_line             0x000007f5
                      DW_AT_external              yes
< 2><0x0000007c>      DW_TAG_formal_parameter
                        DW_AT_name                  v
                        DW_AT_variable_parameter    no
                        DW_AT_type                  <.debug_info+0x000928d2>
< 2><0x00000084>      DW_TAG_null

```

File: src/read/macho/file.rs [hunk 2]
```diff
@@ -287,32 +287,26 @@ where
...
-                            })
-                            .is_some()
-                })
-                .unwrap_or(false)
+        let underscores_suffix = make_suffix(b"__");
+        let underscores_z_suffix = make_suffix(b"__z");
+        let name_matches = |name: &[u8]| {
+            name == section_name
+                || name.strip_prefix(b"__") == underscores_suffix
+                || name.strip_prefix(b"__z") == underscores_z_suffix
```
Comment (philipc) [inline src/read/macho/file.rs:306]: This code looks a lot nicer than what we had before. However, these two comparisons will incorrectly match if both sides are `None` (e.g. this `__z` check will result in matching `__text` with `__data`).

File: src/read/macho/file.rs [hunk 3]
```diff
@@ -287,32 +287,26 @@ where
...
-                    section_name == name
-                        || system_name
-                            .filter(|system_name| {
-                                name.starts_with(b"__") && name[2..] == **system_name
-                            })
-                            .is_some()
-                })
-                .unwrap_or(false)
+        let underscores_suffix = make_suffix(b"__");
+        let underscores_z_suffix = make_suffix(b"__z");
```
Comment (philipc) [inline src/read/macho/file.rs:302]: I'm inclined to be more specific here and match only on `__zdebug_`.

File: tests/read/macho.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#[test]
+fn test_go_macho() {
+    let macho_testfiles = std::path::Path::new("testfiles/macho");
+
+    // Section names we expect to find, whether they should be
+    // compressed, and the actual name of the section in the file.
+    const EXPECTED: &[(&str, bool, &str)] = &[
+        (".debug_abbrev", true, "__zdebug_abbrev"),
+        (".debug_gdb_scripts", false, "__debug_gdb_scri"),
+        (".debug_ranges", true, "__zdebug_ranges"),
```
Comment (philipc) [inline tests/read/macho.rs:14]: Add `("__data", false, "__data"),` here to check for the problem I mention above.

File: src/read/macho/section.rs [hunk 5]
```diff
@@ -102,6 +102,44 @@ where
             .data(self.file.endian, self.internal.data)
             .read_error("Invalid Mach-O section size or offset")
     }
+
+    // Try Go-style "ZLIB" header decompression. This is similar to the GNU-style
+    // format, but the header uses a slightly different format with a 4-byte "ZLIB"
+    // magic followed by a 64-bit big-endian uncompressed size.
```
Comment (philipc) [inline src/read/macho/section.rs:108]: The format looks identical to the GNU-style: GNU use a 4 byte ZLIB magic followed by a 64 bit size, but in the ELF parser we have chosen to limit the size to 32 bits, which we do by checking that the upper 4 bytes of the size are zero.

It'd be nice for both of these formats to handle this the same way and share this code.

**Review by philipc — CHANGES_REQUESTED** [review]

File: tests/read/macho.rs (hunk 4 above)
Comment (ajwerner) [reply tests/read/macho.rs:14]: Done.

File: src/read/macho/section.rs (hunk 5 above)
Comment (ajwerner) [reply src/read/macho/section.rs]: Oh! Now I feel silly, I saw elsewhere the GNU-style thing described as 4 bytes of padding followed by a big-endian u32 ... which literally is a big-endian u64 that fits in u32. My bad.

I've extracted the logic from the elf module into a shared module and used it in both places. A downside I couldn't figure out how to fix without doing something drastic was that I had to strip the section type information from the error messages to reduce the duplication.

If you'd prefer I just keep two copies of the logic around, that's fine. Happy to go whatever route you think is best.

File: src/read/macho/file.rs (hunk 2 above)
Comment (ajwerner) [reply src/read/macho/file.rs]: Good catch! I've found a formulation I like more than before that handles this case. See how you feel about it.

**Review by ajwerner — COMMENTED** [review]
TFTR!

File: src/read/gnu_compression.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        .read::<U32Bytes<_>>(&mut offset)
+        .read_error("GNU compressed section is too short")?
+        .get(endian::BigEndian)
+        .into();
+    let compressed_size = section_size
+        .checked_sub(offset)
+        .read_error("GNU compressed section is too short")?;
+    Ok(Some(CompressedFileRange {
+        format: CompressionFormat::Zlib,
+        offset,
```
Comment (philipc) [inline src/read/gnu_compression.rs:45]: This offset is wrong. Fix this by passing `self.file.data` into this function and use it with `section_offset` instead of using `section.data()`.

Also possibly see if you can avoid passing in `section` at all.

File: src/read/gnu_compression.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    let data = section.data()?;
+    // Assume ZLIB-style uncompressed data is no more than 4GB to avoid accidentally
+    // huge allocations. This also reduces the chance of accidentally matching on a
+    // .debug_str that happens to start with "ZLIB".
+    if data
+        .read_bytes(&mut offset, 8)
+        .read_error("GNU compressed section is too short")?
+        != b"ZLIB\0\0\0\0"
+    {
+        return Err(Error("Invalid ELF GNU compressed section header"));
```
Comment (philipc) [inline src/read/gnu_compression.rs:33]:
```suggestion
        return Err(Error("Invalid GNU compressed section header"));
```

File: src/read/macho/file.rs [hunk 8]
```diff
@@ -287,32 +287,29 @@ where
...
+            let max_len = MAX_SECTION_NAME_LEN - header.len();
+            let suffix = base_name
+                .map(|name| name.get(..max_len).unwrap_or(name))
+                .and_then(|name| name.strip_prefix(prefix.strip_prefix(header).unwrap()));
+            move |name: &[u8]| suffix.is_some() && name.strip_prefix(prefix) == suffix
         };
-
-        self.sections().find(cmp_section_name)
+        let matches_underscores_prefix = make_prefix_matcher(b"__", b"__");
+        let matches_zdebug_prefix = make_prefix_matcher(b"__z", b"__zdebug_");
```
Comment (philipc) [inline src/read/macho/file.rs:305]:
```suggestion
        let matches_underscores_prefix = make_prefix_matcher(b".", b"__");
        let matches_zdebug_prefix = make_prefix_matcher(b".debug_", b"__zdebug_");
```
This version of the code looks correct, but it's a little hard to follow. Can you see if changing `make_prefix_matcher` to accept arguments like this makes it simpler. (So with this change there would be no need for `base_name`.

File: src/read/gnu_compression.rs [hunk 9]
```diff
@@ -0,0 +1,49 @@
+use crate::read::{self, Error, ReadError as _};
+use crate::{endian, CompressedFileRange, CompressionFormat, ReadRef as _, U32Bytes};
+
+// Attempt to parse the the CompressedFileRange for a section using the GNU-style
+// inline compression header format. This is used by the Go compiler in Mach-O files
+// as well as by the GNU linker in some ELF files. This function will return Ok(None)
+// if the section does not have the specified name prefix.
+pub(super) fn maybe_compressed_file_range<'data>(
```
Comment (philipc) [inline src/read/gnu_compression.rs:8]: This will need `#[allow(unused)]` to avoid warnings with the elf and macho features disabled. Use `cargo xtask features` to check this.

**Review by philipc — CHANGES_REQUESTED** [review]

File: src/read/macho/section.rs (hunk 5 above)
Comment (philipc) [reply src/read/macho/section.rs]: The error message changes are fine.

File: src/read/gnu_compression.rs (hunk 9 above)
Comment (ajwerner) [reply src/read/gnu_compression.rs]: I instead marked this whole module as only existing when using macho or elf.

File: src/read/gnu_compression.rs (hunk 7 above)
Comment (ajwerner) [reply src/read/gnu_compression.rs]: Done.

File: src/read/gnu_compression.rs (hunk 6 above)
Comment (ajwerner) [reply src/read/gnu_compression.rs:32]: Done, I moved the section name processing back up to the caller and got rid of the whole "maybe" aspect. See if it pleases.

File: src/read/macho/file.rs (hunk 8 above)
Comment (ajwerner) [reply src/read/macho/file.rs]: Thanks for the suggestion!

**Review by ajwerner — COMMENTED** [review]
Thanks for the review!

**Review by philipc — CHANGES_REQUESTED** [review]
I think this is good to merge after a `cargo fmt`.

**Comment by ajwerner** [comment]
> I think this is good to merge after a cargo fmt.

Sorry for the delay, done. Thank you for the reviews.

**Review by philipc — APPROVED** [review]
